server/src/main/java/org/opensearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -666,8 +669,12 @@ protected void handleReplicaRequest(
             releasable::close
         );
 
+        final ShardId shardId = replicaRequest.getRequest().shardId();
+        assert shardId != null : "request shardId must be set";
+        IndexShard replica = getIndexShard(shardId);
+
         try {
-            new AsyncReplicaAction(replicaRequest, listener, (ReplicationTask) task).run();
+            REPLICA_ACTION_FACTORY.create(replica, getReplicationMode(replica), replicaRequest, listener, (ReplicationTask) task).run();
```

Comment:
I guess we have an issue here, the check is missing `shardRouting.primary()` check 
https://github.com/opensearch-project/OpenSearch/blob/431bdeb1bb17090206ed2e26a486b0ddd00f08bb/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L3340-L3342

which should change how we do `getReplicationMode(replica)`
https://github.com/opensearch-project/OpenSearch/blob/106966018ce57ae4738b278a09601de8737e99af/server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java#L199-L204

Reply:
@Bukhtawar The `shardRouting.primary()` check is added in `ReplicaActionFactory.create` (https://github.com/opensearch-project/OpenSearch/pull/5490/files#diff-58cc23103f9ac2c53091672e90ee1c13fa200f84978a2207bb9acb234cc2d1f1R716)

Reply:
@Bukhtawar  With IndexShard#isRemoteTranslogEnabled, we only want to check if remote translog is enabled or not. The check for whether this is a primary or not has been handled separately. For transport replication actions, we want to know if the remote translog is enabled or not to know what is the replication mode which is present in `TransportReplicationAction`.

Reply:
@Bukhtawar  addressed it.